core/primitives/src/utils.rs

```diff
@@ -279,6 +280,33 @@ pub fn generate_random_string(len: usize) -> String {
     thread_rng().sample_iter(&Alphanumeric).take(len).collect::<String>()
 }
 
+pub struct Serializable<'a, T> {
+    object: &'a T,
+}
+
+impl<'a, T> fmt::Display for Serializable<'a, T>
+where
+    T: serde::Serialize,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", serde_json::to_string(&self.object).unwrap())
+    }
+}
+
+/// Wrap an object that implements Serialize into another object
+/// that implements Display. When used display in this object
+/// it shows its json representation. It is used to display complex
+/// objects using tracing.
+///
+/// ```
+/// tracing::debug!(target: "diagnostic", value=%ser(&object));
+/// ```
+pub fn ser<'a, T>(object: &'a T) -> Serializable<'a, T>
+where
+    T: serde::Serialize,
+{
+    Serializable { object }
```

Comment:
Why do you need to wrap this object instead of serializing it directly?

Reply:
We don't want to pay for what we don't use, serializing by default is very expensive so we only want to serialize if diagnostics are enabled. This pattern allows exactly that by using:
```rust
tracing::debug!(target: "diagnostic", value=%ser(&object));
```
Trying to serialize directly in that line is more verbose, and require additional workaround since you can't pass String object as [tracing_core::field::Value](https://docs.rs/tracing-core/0.1.7/tracing_core/field/trait.Value.html).

Reply:
Nice trick. It is more common though to wrap structs like this: 
```rust
pub struct Serializable<'a, T>(&'a T)
```
which is also known to be optimized out by the compiler.

Reply:
@nearmax addresed in https://github.com/nearprotocol/nearcore/pull/1928/commits/9a46a38c3dae21d174f30013172de8b2c4bbb511